R/attr.R

```diff
@@ -163,25 +163,56 @@ has_name <- function(x, name) {
   name %in% names2(x)
 }
 
-#' Set names of a vector.
+#' Set names of a vector
 #'
-#' This is a snake case wrapper for [stats::setNames()], with tweaked
-#' defaults, and stricter argument checking.
+#' This is equivalent to [stats::setNames()], with more features and
+#' stricter argument checking.
 #'
 #' @param x Vector to name.
-#' @param nm Vector of names, the same length as `x`.
+#' @param nm,... Vector of names, the same length as `x`.
+#'
+#'   You can specify names in three ways:
+#'
+#'   * If you do nothing, `x` will be named with itself.
+#'
+#'   * If `x` already has names, you can provide a function or formula
+#'     to transform the existing names. In that case, `...` is passed
+#'     to the function.
+#'
+#'   * In all other cases, `nm` and `...` are passed to [chr()]. This
+#'     gives implicit splicing semantics: you can pass character
+#'     vectors or list of character vectors indistinctly.
 #' @export
 #' @examples
 #' set_names(1:4, c("a", "b", "c", "d"))
+#' set_names(1:4, letters[1:4])
+#' set_names(1:4, "a", "b", "c", "d")
 #'
 #' # If the second argument is ommitted a vector is named with itself
 #' set_names(letters[1:5])
-set_names <- function(x, nm = x) {
+#'
+#' # Alternatively you can supply a function
+#' set_names(1:10, ~ letters[seq_along(.)])
+#' set_names(head(mtcars), toupper)
+#'
+#' # `...` is passed to the function:
+#' set_names(head(mtcars), paste0, "_foo")
+set_names <- function(x, nm = x, ...) {
   if (!is_vector(x)) {
-    stop("`x` must be a vector", call. = FALSE)
+    abort("`x` must be a vector")
+  }
+
+  if (is_function(nm) || is_formula(nm)) {
+    nm <- as_function(nm)
+    nm <- nm(names2(x), ...)
+  } else if (!is_null(nm)) {
```

Comment:
I think you can drop the `is_null()` stuff.  I'm now not sure why I thought it was necessary